text/0000-futures.md

```diff
@@ -0,0 +1,908 @@
+anticipate follow-up RFCs ultimately bringing most of the library into libcore (to
+provide a complete complement of APIs).
+
+The proposed APIs are based on the futures crate, but with two major changes:
+
+- The use of [pinned types] to enable borrowing within futures.
+- Removing the associated `Error` type (and adjusting combinators accordingly),
+  in favor of just using `Output = Result<T, E>` instead. The RFC includes an
+  extension trait to provide conveniences for `Result`-producing futures as
+  well.
+
+[pinned types]: https://github.com/rust-lang/rfcs/pull/2349
+[companion RFC]: https://github.com/rust-lang/rfcs/pull/2394
+
+# Motivation
+[motivation]: #motivation
+
+There are two reasons to consider bringing futures into the standard library.
+
+The first, and by far most important, is to provide a supporting mechanism for
+`async`/`await` syntax:
+
+```rust
+async fn function(argument: &str) -> usize {
+     // ...
+}
+```
+
+The syntax itself is motivated in the [companion RFC]. As with closures, it
+involves producing an anonymous type, so that the above declaration is
+equivalent to:
+
+```rust
+fn function<'a>(argument: &'a str) -> _Anonymous<'a, usize> {
+     // ...
+}
+```
+
+and, again like a closure, the type is only usable through the trait it
+implements: `Future`. Hence, to include this syntax in the language, we must
+also introduce `Future` into `core` as a lang item.
+
+The second reason to introduce futures is to establish them more formally as
+*the* way to express composable, asynchronous computation in Rust. Over time,
+the futures library has been refined to a core that bakes in almost no
+assumptions and is usable in a wide variety of contexts, including operating
+systems and embedded devices. As such, it has also become a *de facto* standard
+in this space, and there is a strong desire for the core parts of the library to
+reach the same level of stability as `std` itself. Bringing futures into libcore
+is a final step in standardization.
+
+## Changes from futures 0.2
+
+### Pinning
+
+The motivation to use the `Pin` is simple: it makes it possible for futures to
+hold *internal borrows* across yield points, which allows for far more idiomatic
+futures-based APIs. For example, an async `read` method currently has to take
+*ownership* of both the I/O object and the buffer to read into, and then yield
+ownership on success or failure:
+
+```rust
+fn read<T: AsMut<[u8]>>(self, buf: T) ->
+    impl Future<Item = (Self, T, usize), Error = (Self, T, io::Error)>
+```
+
+This design is needed to ensure that the resulting future is `'static`, which is
+in turn needed because borrowing across yield points isn't possible. It's very
+unwieldy.
+
+Once we lift this restriction, however, we can write a signature that is
+literally the `async` version of the usual `read`:
+
+```rust
+async fn read(&mut self, buf: &mut [u8]) -> io::Result<usize>;
```

Comment:
Perhaps a more specific example can be included, one that cannot be implemented today? An async function returning a future that borrows both `self` and `buf` already is possible, [for example](https://play.rust-lang.org/?gist=7807e380c33c2b7a521c1ce7bf15ba20&version=beta) (beta just used for `impl Future`).

Reply:
This specific function can't actually be used today (at least not without lots of unsafe code), since the future return type is bound to the lifetimes of the input types.

Reply:
Hm, can you expand on that? The link shows the function actually is used, being put in the `block_on`. 

Do you mean you can't really return this future from another function, since the TCP stream and buffer would be dropped? That's true, but the same problem exists with `async fn`.